cas-server-support-osf/src/main/java/io/cos/cas/authentication/handler/support/OpenScienceFrameworkPrincipalFromRequestRemoteUserNonInteractiveCredentialsAction.java

```diff
@@ -269,65 +279,102 @@ protected Credential constructCredentialsFromRequest(final RequestContext contex
+
         if (StringUtils.hasText(shibbolethSession)) {
+            // SAML-Shibboleth based institution login
             final String remoteUser = request.getHeader(REMOTE_USER);
             if (StringUtils.isEmpty(remoteUser)) {
                 logger.error("Invalid Remote User specified as Empty");
                 throw new RemoteUserFailedLoginException("Invalid Remote User specified as Empty");
+            } else {
+                logger.info("Remote User from HttpServletRequest '{}'", remoteUser);
+                credential.setRemotePrincipal(Boolean.TRUE);
             }
-
-            logger.info("Remote User from HttpServletRequest '{}'", remoteUser);
-            credential.setRemotePrincipal(Boolean.TRUE);
-
             for (final String headerName : Collections.list(request.getHeaderNames())) {
                 if (headerName.startsWith(ATTRIBUTE_PREFIX)) {
                     final String headerValue = request.getHeader(headerName);
                     logger.debug("Remote User [{}] Auth Header '{}': '{}'", remoteUser, headerName, headerValue);
-
                     credential.getAuthenticationHeaders().put(
-                            headerName.substring(ATTRIBUTE_PREFIX.length()),
-                            headerValue
+                        headerName.substring(ATTRIBUTE_PREFIX.length()),
+                        headerValue
                     );
                 }
             }
-
             // Notify the OSF of the remote principal authentication.
-            final PrincipalAuthenticationResult remoteUserInfo = notifyRemotePrincipalAuthenticated(credential);
+            final PrincipalAuthenticationResult remoteUserInfo
+                = notifyRemotePrincipalAuthenticated(credential, null, PROTOCOL_SAML);
             credential.setUsername(remoteUserInfo.getUsername());
             credential.setInstitutionId(remoteUserInfo.getInstitutionId());
-
+            return credential;
+        } else if (context.getFlowScope().get("authenticationDelegationProtocol") == PROTOCOL_CAS) {
+            // CAS-pac4j based institution login
+            TicketGrantingTicket ticketGrantingTicket;
+            Principal principal;
+            try {
+                final String ticketGrantingTicketId
+                    = (String) context.getFlowScope().get("ticketGrantingTicketId");
+                ticketGrantingTicket = centralAuthenticationService.getTicket(
+                    ticketGrantingTicketId,
+                    TicketGrantingTicket.class
+                );
+            } catch (final InvalidTicketException e) {
+                logger.error("Invalid Ticket Granting Ticket");
+                throw new RemoteUserFailedLoginException("Invalid Ticket Granting Ticket");
+            }
+            try {
+                credential.setRemotePrincipal(Boolean.TRUE);
+                principal = ticketGrantingTicket.getAuthentication().getPrincipal();
```

Comment:
Retrieve the `principal` from `TGT` which contains authenticated institution user's information.